### ISSUE TITLE
Same Server Dependency

### ISSUE THREAD

**Issue by thehabes** [CenterForDigitalHumanities/rerum_server#122]
I noticed that if you have a MongoDB running on a separate server from you rerum_server deployment, you get an error like

`The origin server did not find a current representation for the target resource or is not willing to disclose that one exists`

It has something to do with Constant.java and struts.xml.

How I noticed the error:

Deployed a rerum_server on localhost successfully, loaded index.jsp.
Wanted to use MongoDB on dev server so I didn't have to set up localhost.
Changed `Constant.RERUM_BASE`,`Constant.RERUM_ID_PREFIX`, and `Constant.RERUM_AGENT_CLAIM` to point to dev so the MongoDB objects would be formatted correctly when saved.  
Tried to load index.jsp Got the error.  
Reverted changes, loaded index.jsp again, no more error.

This may be the reason the AWS deployment is having trouble.  We need to investigate to understand EXACTLY WHY this happens.  

At the moment, it seems to be causing a terrible dependency: rerum_server and mongo have to be on the same server.

**Comment by thehabes** [comment]
I am going to add notes here as I find things.

### First pass localhost setup intending to replicate the error for investigation
I have tomcat 8.5 on my localhost.  It is where I am deploying applications.

I am deploying a `rerum_server` as `/v1` whose `Constants` all point at `localhost:8080`.  The bot for this instance is the `sandbox@rerum.io` bot.  The Mongo Client is `MongoClient("165.134.107.94", 27017)`.

I am deploying a TinyThings as `/app/` whose `Constants` all point at `localhost:8080/v1/`.  `localhost:8080/app` uses the `sanbox@rerum.io` properties and has the sandbox agent encoded in the access token it uses for the `localhost:8080/v1/` API processor.  This will make all `CRUD` coming from `localhost:8080/app/` pass authorization in `localhost:8080/v1/`.  

The Tomcat logs from `localhost:8080` say
`
java.io.IOException: http://localhost:8080/app/create/ returned HTTP response code: 500 for URL: http://localhost:8080/v1/api/create.action
`
<html lang="en"><head></head><body><h1>HTTP Status 500 ? Internal Server Error</h1><hr class="line" /><p><b>Type</b> Exception Report</p><p><b>Message</b> Illegal char &lt;:&gt; at index 3: jar:file:\C:\Users\Habes\Documents\tomcat\webapps\v1\WEB-INF\lib\struts2-core-2.3.12.jar</p><p><b>Description</b> The server encountered an unexpected condition that prevented it from fulfilling the request.</p><p><b>Exception</b></p><pre>java.nio.file.InvalidPathException: Illegal char &lt;:&gt; at index 3: jar:file:\C:\Users\Habes\Documents\tomcat\webapps\v1\WEB-INF\lib\struts2-core-2.3.12.jar     java.base&#47;sun.nio.fs.WindowsPathParser.normalize(Unknown Source)    java.base&#47;sun.nio.fs.WindowsPathParser.parse(Unknown Source)    java.base&#47;sun.nio.fs.WindowsPathParser.parse(Unknown Source)        java.base&#47;sun.nio.fs.WindowsPath.parse(Unknown Source)  java.base&#47;sun.nio.fs.WindowsFileSystem.getPath(Unknown Source)      java.base&#47;java.io.File.toPath(Unknown Source)       java.base&#47;java.util.zip.ZipFile$Source.get(Unknown Source)      java.base&#47;java.util.zip.ZipFile$CleanableResource.&lt;init&gt;(Unknown Source)      java.base&#47;java.util.zip.ZipFile$CleanableResource.get(Unknown Source)   java.base&#47;java.util.zip.ZipFile.&lt;init&gt;(Unknown Source)        java.base&#47;java.util.zip.ZipFile.&lt;init&gt;(Unknown Source)    java.base&#47;java.util.jar.JarFile.&lt;init&gt;(Unknown Source)        java.base&#47;java.util.jar.JarFile.&lt;init&gt;(Unknown Source)    java.base&#47;java.util.jar.JarFile.&lt;init&gt;(Unknown Source)        com.opensymphony.xwork2.util.fs.JarEntryRevision.needsReloading(JarEntryRevision.java:76)   com.opensymphony.xwork2.util.fs.DefaultFileManager.fileNeedsReloading(DefaultFileManager.java:66)       com.opensymphony.xwork2.config.providers.XmlConfigurationProvider.needsReload(XmlConfigurationProvider.java:394)    org.apache.struts2.config.StrutsXmlConfigurationProvider.needsReload(StrutsXmlConfigurationProvider.java:169)       com.opensymphony.xwork2.config.ConfigurationManager.needReloadContainerProviders(ConfigurationManager.java:215)     com.opensymphony.xwork2.config.ConfigurationManager.conditionalReload(ConfigurationManager.java:179)    com.opensymphony.xwork2.config.ConfigurationManager.getConfiguration(ConfigurationManager.java:73)  org.apache.struts2.dispatcher.Dispatcher.getContainer(Dispatcher.java:942)
        org.apache.struts2.dispatcher.ng.PrepareOperations.createActionContext(PrepareOperations.java:77)       org.apache.struts2.dispatcher.ng.filter.StrutsPrepareAndExecuteFilter.doFilter(StrutsPrepareAndExecuteFilter.java:78)       org.tuckey.web.filters.urlrewrite.RuleChain.handleRewrite(RuleChain.java:176)       org.tuckey.web.filters.urlrewrite.RuleChain.doRules(RuleChain.java:145) org.tuckey.web.filters.urlrewrite.UrlRewriter.processRequest(UrlRewriter.java:92)   org.tuckey.web.filters.urlrewrite.UrlRewriteFilter.doFilter(UrlRewriteFilter.java:394)  org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:88)    org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:76)</pre></body></html>

In `localhost:8080/app/create/` (`TinySave.java`) this line reading the response from `localhost:8080/v1/` is failing `BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(),"utf-8"));` is failing because the connection has status `500`.  

Given this evidence, I believe the error is located in `http://localhost:8080/v1/`.  The first line of ObjectAction.java#saveNewObject is to log the string `"create object"`.  This string is not appearing in my `localhost:8080` tomcat logs, which means the request from `localhost:8080/app/create/` isn't even making through servlet mapping.  This could be because ObjectAction.java is failing to initialize one of its globals, like the `mongoDBService`.  I will continue to investigate by putting `System.out()` deeper in the back end, especially around `MongoClient`, `MongoDBService` and `MongoDBUtil`.

**Comment by thehabes** [comment]
The deploys are failing in our localhost windows tomcat servers because of https://issues.apache.org/jira/browse/WW-4970

**Comment by thehabes** [comment]
Thanks to some co-op with @snsarma we know what the issue is.  It comes from `MongoUtil.java`.

We deployed this on dev-01 and we got a `Mongo Connection Refused` error.

This led us to determine the firewall is blocking access to the running `mongod` service from all incoming connections.  When we replaced that line with a client pointing at localhost instead

The deployment worked.

This is, in fact, is a same server dependency.  If we want the mongo on dev to be allowed as a server-to-server connection, we probably have to open up the mongo port.
